pkg/controller/ecr/repositorypolicy/controller.go

```diff
@@ -0,0 +1,209 @@
+	ecr "github.com/crossplane/provider-aws/pkg/clients/ecr"
+)
+
+const (
+	errUnexpectedObject = "managed resource is not an repository resource"
+
+	errCreate       = "failed to create repository policy"
+	errGet          = "failed to get repository policy"
+	errUpdate       = "failed to update repository policy"
+	errDelete       = "failed to delete the repository resource"
+	errSpecUpdate   = "cannot update spec of repository policy custom resource"
+	errStatusUpdate = "cannot update status of repository policy custom resource"
+	errNotSpecified = "failed to format Repository Policy, no rawPolicy or policy specified"
+)
+
+// SetupRepositoryPolicy adds a controller that reconciles ECR.
+func SetupRepositoryPolicy(mgr ctrl.Manager, l logging.Logger, rl workqueue.RateLimiter) error {
+	name := managed.ControllerName(v1alpha1.RepositoryPolicyGroupKind)
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		WithOptions(controller.Options{
+			RateLimiter: ratelimiter.NewDefaultManagedRateLimiter(rl),
+		}).
+		For(&v1alpha1.RepositoryPolicy{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(v1alpha1.RepositoryPolicyGroupVersionKind),
+			managed.WithExternalConnecter(&connector{kube: mgr.GetClient()}),
+			managed.WithReferenceResolver(managed.NewAPISimpleReferenceResolver(mgr.GetClient())),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type connector struct {
+	kube client.Client
+}
+
+func (c *connector) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	cr, ok := mg.(*v1alpha1.RepositoryPolicy)
+	if !ok {
+		return nil, errors.New(errUnexpectedObject)
+	}
+	cfg, err := awsclient.GetConfig(ctx, c.kube, mg, cr.Spec.ForProvider.Region)
+	if err != nil {
+		return nil, err
+	}
+	return &external{client: awsecr.New(*cfg), kube: c.kube}, nil
+}
+
+type external struct {
+	kube   client.Client
+	client ecr.RepositoryPolicyClient
+}
+
+func (e *external) Observe(ctx context.Context, mgd resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mgd.(*v1alpha1.RepositoryPolicy)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errUnexpectedObject)
+	}
+
+	response, err := e.client.GetRepositoryPolicyRequest(&awsecr.GetRepositoryPolicyInput{
+		RegistryId:     cr.Spec.ForProvider.RegistryID,
+		RepositoryName: cr.Spec.ForProvider.RepositoryName,
+	}).Send(ctx)
+
+	if err != nil {
+		if ecr.IsRepoNotFoundErr(err) {
+			return managed.ExternalObservation{}, nil
+		}
+		return managed.ExternalObservation{}, awsclient.Wrap(resource.Ignore(ecr.IsPolicyNotFoundErr, err), errGet)
+	}
+
+	policyData, err := e.formatRepositoryPolicy(cr)
+
+	if err != nil {
+		return managed.ExternalObservation{}, awsclient.Wrap(resource.Ignore(ecr.IsPolicyNotFoundErr, err), errGet)
+	}
+	cr.Status.AtProvider.PolicyText = *policyData
+
+	if !cmp.Equal(*policyData, *response.PolicyText) {
```

Comment:
thats my understanding.

Reply:
@enderv I think are some places where we do this with the assumption that it will never be `nil`. I personally prefer to check that it is not `nil` as, though it wouldn't be great by SDK maintainers to change this behavior, it could be changed and we would then have a potential panic here. What do you think?

Reply:
sounds good to me. On it

Reply:
Oh actually already got this covered with my lateinitialization change have a nil check in the function there and now just comparing the `RepositoryPolicyParameters`

Reply:
Aren't we still deferencing before late init?

https://github.com/crossplane/provider-aws/pull/573/files#diff-b54d7ac2e4ab0e950fd8fb8c791b11196da72121140174398d12c779bc643624R116

Reply:
:eyes: so we are. on it.